xds/src/test/java/io/grpc/xds/FakeControlPlaneXdsIntegrationTest.java

```diff
@@ -0,0 +1,332 @@
+import io.grpc.testing.protobuf.SimpleResponse;
+import io.grpc.testing.protobuf.SimpleServiceGrpc;
+import java.net.ServerSocket;
+import java.util.Collections;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/**
+ * Xds integration tests using a local control plane, implemented in
+ * {@link XdsTestControlPlaneService}.
+ * Test cases can inject xds configs to the control plane for testing.
+ */
+@RunWith(JUnit4.class)
+public class FakeControlPlaneXdsIntegrationTest {
+  private static final Logger logger =
+      Logger.getLogger(FakeControlPlaneXdsIntegrationTest.class.getName());
+
+  protected int testServerPort;
+  protected int controlPlaneServicePort;
+  private Server server;
+  private Server controlPlane;
+  private XdsTestControlPlaneService controlPlaneService;
+  protected SimpleServiceGrpc.SimpleServiceBlockingStub blockingStub;
+  private XdsNameResolverProvider nameResolverProvider;
+  private static final String scheme = "test-xds";
+  private static final String SERVER_LISTENER_TEMPLATE =
+      "grpc/server?udpa.resource.listening_address=%s";
+  private static final String rdsName = "route-config.googleapis.com";
+  private static final String clusterName = "cluster0";
+  private static final String edsName = "eds-service-0";
+  private static final String HTTP_CONNECTION_MANAGER_TYPE_URL =
+      "type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3"
+          + ".HttpConnectionManager";
+
+  /**
+   * For test purpose, use boostrapOverride to programmatically provide bootstrap info.
+   */
+  private Map<String, ?> defaultBootstrapOverride() {
+    return ImmutableMap.of(
+        "node", ImmutableMap.of(
+            "id", UUID.randomUUID().toString(),
+            "cluster", "cluster0"),
+        "xds_servers", Collections.singletonList(
+
+            ImmutableMap.of(
+                "server_uri", "localhost:" + controlPlaneServicePort,
+                "channel_creds", Collections.singletonList(
+                    ImmutableMap.of("type", "insecure")
+                ),
+                "server_features", Collections.singletonList("xds_v3")
+            )
+        ),
+        "server_listener_resource_name_template", SERVER_LISTENER_TEMPLATE
+    );
+  }
+
+  /**
+   * 1. Start control plane server and get control plane port
+   * 2. Find an available port for xds server
+   * 3. Set control plane config using the port in 2
+   * 4. Start xds server using control plane port in 1 and available port in 2
+   * */
+  @Before
+  public void setUp() throws Exception {
+    startControlPlane();
+    ServerSocket serverSocket = new ServerSocket(0);
+    testServerPort = serverSocket.getLocalPort();
+    serverSocket.close();
```

Comment:
I just noticed how you're choosing the port here. That sorta works if you are starting a server once on a machine doing nothing else. But in tests, it is pretty unreliable. It'll probably cause flakes.

I considered several solutions (test function to change the listener name in XdsServer, support wildcard names in the fake control plane), but actually all we have to do is remove the `%s` from `SERVER_LISTENER_TEMPLATE`. If you do that, then we can predict the resource name and it looks like we can use `XdsServerBuilder.forPort(0, ...)`. If that doesn't work out, we'll need to consider some other solution as a follow-up.

Reply:
Aha! Very interesting problem and nice solution!
If we only `setXdsConfig` at the control plane once as is now, it won't work, because the eds name and rds domain needs the socket address that needs port in order for the client and server to communicate. However, we can first use `SERVER_LISTENER_TEMPLATE` without `%s` to let the server run, using `XdsServerBuilder.forPort(0, ...)`. Then we get the server port and call `setXdsConfig` to provide the real port.
However, I see there is a potential flakiness due to the race between the client sends request and the new version of xds config has not been propagated from the control plane to the client and server...

Reply:
Hmm... aren't those separate resources though? We can initialize the server's resources, let it start. Then we have a port and can initialize the client's resources and let it start. So yes, we'll do `setXdsConfig()` twice, but as long as we start the client last, we're okay and there's no race.

Reply:
See the latest update. Verified this:
Used a fake authority for server listener to start, it does not matter at this time. Then we get the real port to update LDS for the server `httpConnectionManager` authority, used in [server routing config](https://github.com/grpc/grpc-java/blob/a1c41e3d30b7621b0fa626107013597568c189e2/xds/src/main/java/io/grpc/xds/XdsServerWrapper.java#L758), but technically we can also use `overrideAuthority()` in client channel to let them match with a random authority for test, right?

Reply:
The xds name doesn't have to relate in any way to reality. It seems you should just use `xds:///test-server` on client-side and register the ApiListener with the same name. EDS is the only thing that needs a real IP+port.

Reply:
hm, I was actually saying that the domain/virtual host match at the server side. The domain is calculated by the server  call authority and it has to match the rds configuration we put in the control plane config.
https://github.com/grpc/grpc-java/blob/fbb1dbf7a519671331decf5ab3e0ed9716e1823a/xds/src/main/java/io/grpc/xds/XdsServerWrapper.java#L758

Reply:
> The domain is calculated by the server call authority

The ServerCall authority comes from the client. If the client specified `xds:///test-server` then the authority seen on server-side is `test-server`. In addition, you can use `*` for the virtual host match, and then the authority doesn't matter.